agent/utils/retry/retry.go

```diff
@@ -56,6 +63,91 @@ func RetryWithBackoffCtx(ctx context.Context, backoff Backoff, fn func() error)
 	}
 }
 
+func RetryWithBackoffCtxForTaskHandler(ctx context.Context, eventFlowCtx context.Context, cfg *config.Config, taskARN string, backoff Backoff, delay time.Duration, fn func() error) error {
+
+	var err error
+	for {
+		select {
+		case <-ctx.Done():
+			return nil
+		default:
+		}
+
+		err = fn()
+
+		retriableErr, isRetriableErr := err.(apierrors.Retriable)
+
+		//if unretriable error in disconnected mode, don't return
+		if err == nil || (!cfg.GetDisconnectModeEnabled() && isRetriableErr && !retriableErr.Retry()) {
+			return err
+		}
+
+		/*
+			If we were in normal mode up to this point and switch to disconnected mode here,
+			we enter the if block
+
+			If we were in disconnected mode up to this point and switch to normal mode here,
+			eventFlowCtx is cancelled, and we enter else block
+		*/
+		if cfg.GetDisconnectModeEnabled() && eventFlowCtx != nil {
+
+			/*
+				If we switch to normal mode while we are in this portion before calling WaitForDurationAndInterruptIfRequired,
+				there is a chance that we receive on a cancelled context- but this returns immediately
+			*/
+			WaitForDurationAndInterruptIfRequired(eventFlowCtx, delay)
+		} else {
+
+			/*
+				If we switch to disconnected mode at this point, eventFlowCtx is initialized, no effect
+			*/
+
+			waitForDuration(backoff.Duration())
+		}
+
+		/*
+			If we were in disconnected mode up to this point and switch to normal mode here,
+			eventFlowCtx context is cancelled, but this has no effect
+
+			If we were in normal mode up to this point and switch to disconnected mode here,
+			eventFlowCtx is initialzed, but this has no effect
+		*/
+
+	}
+}
+
+func waitForDuration(delay time.Duration) bool {
+	reconnectTimer := time.NewTimer(delay)
+	select {
+	case <-reconnectTimer.C:
+		return true
+
+	default:
+		return false
+	}
+}
+
+func WaitForDurationAndInterruptIfRequired(eventFlowCtx context.Context, delay time.Duration) bool {
```

Comment:
nit: the more idiomatic name for this would be `waitForDurationWithContext` (see https://go.dev/blog/context-and-structs)